### ISSUE TITLE
Do we need application specific token?

### ISSUE THREAD

**Issue by rinigus** [poetaster/keypeer.org#2]
I would argue against using application specific tokens. In my mind, keypeer should just provide API keys on application request that follows some REST API. Within that request, user's token and requested API key identifier should be used. For example:

https://keypeer.org/get?token=MySecretUserToken&app=mapbox.com

I don't see why would it be interested to know which application was requesting API key.

**Comment by rinigus** [comment]
Maybe you were meaning the same, but I understood it wrong from the text and the schemes.

**Comment by poetaster** [comment]
It basically boils down to the same thing. my idea of hashing User part and App part together was just an idea. It could certainly be as you suggest!

**Comment by poetaster** [comment]
> I don't see why would it be interested to know which application was requesting API key.

One of the points you raise in #4 and the encryption scheme I have in mind for securing the data suggest having multiple keys. It depends a lot on whether users re-use a 'user' key.  In that case we need to have an application or api provider specific key even if it was just in the sense of a key in a database table. But I'd take that one step further. I'm thinking in the sense of public and private keys for proxy re-encryption .

**Comment by rinigus** [comment]
I don't follow your argument, sorry. If you mean that open payment token would be derived from the main API key of the user then it still doesn't bind app and the key. Maybe I misunderstood something, so let me bring an example. Is there any reason why user cannot use the same keypeer API key for Pure Maps and some other map application? 

As for whether the "user" key is re-used - I presume we talk about keypeer API key, I think it is OK to reuse. There is always an alternative to generate a new one. But again, it is a hassle. Way easier is to use the same key and token for next payment.

Argument with proxy I don't follow. Do you want to provide proxy service between app and the end-service? That would make it more complicated and maybe incompatible with services. I have a feeling that there are some services which don't want you to use some central proxy and ask to have direct contact between app and service. But I am not sure whether it is true as I cannot immediately find it.

Rather long reply to your short one. Feel free to ignore if it is not important.

**Comment by poetaster** [comment]
I think you're thinking of 'user makes an account on keypeer and attaches various services' to it or something like that?

I'm thinking of not re-using user keys because it's just not secure and it makes one app subject to possible leaks in another. That's why I wanted to combine application and user keys. But this can effectively be fleshed out later. 

We are effectively a proxy between an api provider (which publishes keys) and an end user application, a subscriber.   The question of how you safely keep a data store encrypted when you have multiple publishers and subscribers 'sharing' key information has been studied in depth. We fall into the category of proxy (peer).  The academic background: https://eprint.iacr.org/2017/410 You can ignore most of the math and focus on the workflow ideas.

But the key exchange/sharing stuff is only one part of the issue. I'm trying to keep user data to a bare minimum and that's where app instance stands in for the user. I'll make some more diagrams :)

**Comment by rinigus** [comment]
I'll have to look into the paper, replying now to describe my intention.

As such, no account is really needed. User can "register" by visiting a page that, without asking anything, will:

- generate random API key. That's gonna be used as user's secret
- Use keypeer secret key (only keypeer knows it) to sign the user's API key
- signature of the API key is user's token
- User is shown user's API key and token

User can use API key to communicate with keypeer and ask for service API keys. User will use token as a tag for the payments.

On keypeer side, keypeer can just take API key from the request, sign it and compare with the payed tokens. If token is in good standing, keypeer will return requested service API key.

With such scheme, user can share the keypeer key between apps.

Now, I don't know whether this scheme is reasonably secure assuming that tokens are visible to all via public payment system. It may require changing keypeer secret at some interval to avoid using the same key too many times. Which is not a big deal, as API keys can have prefixes indicating a key that was used to sign it to get a token.

However, as you could see, we will not get any user data via such scheme. Just if user will loose API key, all payments/donations are lost as well.

**Comment by poetaster** [comment]
Yeah, now you're describing more or less what I had in mind :) I obviously didn't do a good job of communicating that. 

I was thinking of the initial key generation being done in the application  with a specific call, in app. But you're going in the direction I was thinking.

> Now, I don't know whether this scheme is reasonably secure assuming that tokens are visible to all via public payment system.

I think it's a public/private key like distinction. We don't want the private part being published ;) But we have to contact payment providers anyway ....

> However, as you could see, we will not get any user data via such scheme. Just if user will loose API key, all payments/donations are lost as well.

A number of services that use, for instance, 2FA have recovery schemes that could be implemented.  Usually it's a custom login url with a backup key (hetzner for instance has an account id login with a recovery key for users of yubikeys). Sometimes it's a list of several keys that are used on recovery forms in multiple steps (proxmox). We'll figure something out.

**Comment by poetaster** [comment]
What speaks for application keys:

* ease usage stats (mapbox api key used by multiple apps, but uneven distribution of requests)
* ability to block individual apps without blocking a users other use of their key

**Comment by poetaster** [comment]
ah, you had more or less the same scheme for key generation in #4 under Tokens. I'm going to copy some of these schemes into workflow documents now.

**Comment by rinigus** [comment]
I can see that the key generation and usage flow is now in the text, thanks!

As an app writer, I would prefer to keep apps away from registration of API key and token. Otherwise we would have to re-implement it in many apps and follow that API as well. Way easier is to let it be done in one place - centrally. Apps would have to just communicate using keypeer API key to get the keys that they are interested in.

Re private key: using the scheme that I described, private stays with keypeer only. Not needed for user and payment provider.

Re recovery: simpler option is to suggest making small donations and take care of the keys. For example, Bitwarden cannot do anything if you loose your master password - all your data will be kept encrypted. They warn about it and that's all. Same here: we can just warn about it and let it be. Note that we can provide new token on the basis of API key, but not restore API key. Although, not sure even token regeneration is needed ...

**Comment by rinigus** [comment]
Re blocking apps: not sure why do we want it. In the end, it is a user that is either ON or OFF.

Re stats: I think apps can request only keys that they are interested in. And request only when needed. Thus, Pure Maps will request mapbox only when user wants to use it. That way, it will not be marked as used by that user until it is really needed.

Obviously, after request it will be cached - for as long as specified by keypeer in response. That way, we don't need repetitive requests.
